src/util/CancellationHandle.h

```diff
@@ -37,10 +68,44 @@ static_assert(std::atomic<CancellationState>::is_always_lock_free);
 
 /// Thread safe wrapper around an atomic variable, providing efficient
 /// checks for cancellation across threads.
+template <bool WatchDogEnabled = areExpensiveChecksEnabled>
 class CancellationHandle {
   std::atomic<CancellationState> cancellationState_ =
       CancellationState::NOT_CANCELLED;
 
+  template <typename T>
+  using WatchDogOnly = std::conditional_t<WatchDogEnabled, T, detail::Empty>;
+  [[no_unique_address]] WatchDogOnly<std::atomic_bool> watchDogRunning_{false};
+  [[no_unique_address]] WatchDogOnly<ad_utility::JThread> watchDogThread_;
+  [[no_unique_address]] WatchDogOnly<std::chrono::steady_clock::rep>
+      startTimeoutWindow_{0};
+
+  template <typename... ArgTypes>
+  void pleaseWatchDog(CancellationState state,
+                      const InvocableWithConvertibleReturnType<
+                          std::string_view, ArgTypes...> auto& detailSupplier,
+                      ArgTypes&&... argTypes) requires WatchDogEnabled {
+    AD_CORRECTNESS_CHECK(watchDogRunning_.load(std::memory_order_relaxed));
```

Comment:
I understand. But that basically means that you always need to start the watchdog before the first call to checkTimeout. So that is even more a hint for starting it in the constructor. Or you just turn this into a if(...). 
And for this to work, you have to use `release` semantics AFTER you have set up the thread technically (or just leave out the check again).

Reply:
But then again you don't access the watchdog, so everything should be fine, but that check doesn't do too much.